libvips/iofuncs/threadpool.c

```diff
@@ -538,251 +221,308 @@ vips_thread_state_new( VipsImage *im, void *a )
 		VIPS_TYPE_THREAD_STATE, vips_thread_state_set, im, a ) ) );
 }
 
-/* A VipsTask is the state of one call to vips_threadpool_run().
+/* What we track for each thread in the pool.
  */
-typedef struct _VipsTask {
-	/* All private.
-	 */
+typedef struct _VipsWorker {
+	/*< private >*/
+	struct _VipsThreadpool *pool; /* Pool we are part of */
+
+	VipsThreadState *state;
+
+        gboolean stop;
+
+} VipsWorker;
+
+/* What we track for a group of threads working together.
+ */
+typedef struct _VipsThreadpool {
 	/*< private >*/
 	VipsImage *im;		/* Image we are calculating */
 
-	/* Start or reuse a thread, do a unit of work (runs in parallel) 
-	 * and allocate a unit of work (serial). Plus the mutex we use to 
-	 * serialize work allocation.
+	/* Start a thread, do a unit of work (runs in parallel) and allocate 
+	 * a unit of work (serial). Plus the mutex we use to serialize work 
+	 * allocation.
 	 */
 	VipsThreadStartFn start; 
 	VipsThreadpoolAllocateFn allocate;
 	VipsThreadpoolWorkFn work;
 	GMutex *allocate_lock;
         void *a; 		/* User argument to start / allocate / etc. */
 
-	/* The caller blocks here until all tasks finish.
+	int max_workers;	/* Max number of workers in pool */
+
+	/* The number of workers in the pool (as a negative number, so
+         * -4 means 4 workers are running).
 	 */
-	VipsSemaphore finish;	
+	VipsSemaphore n_workers;
 
 	/* Workers up this for every loop to make the main thread tick.
 	 */
-	VipsSemaphore tick;	
+	VipsSemaphore tick;
+
+	/* The number of workers queueing up on allocate_lock. Use this to
+         * grow and shrink the threadpool.
+	 */
+	int n_waiting;
 
 	/* Set this to abort evaluation early with an error.
 	 */
 	gboolean error;		
 
-	/* Set by Allocate (via an arg) to indicate normal end of computation.
+	/* Ask threads to exit, either set by allocate, or on free.
 	 */
 	gboolean stop;
-} VipsTask;
 
-/* Allocate some work (single-threaded), then do it (many-threaded).
+        /* Set this and the next worker to see it will clear the flag and exit
+         * (used to downsize the threadpool).
+         */
+        int exit;
```

Comment:
nit: the newly added files and this one contains mixed indentations (4-space versus tabs). Perhaps we should re-indent?

Reply:
Good point, I've been fiddling with my vim settings and I've probably messed it up.

Perhaps we should do tabs everywhere (I don't really mind). Do you have a preference?

Reply:
I don't have a preference, as long as it's consistent. Tabs are the most commonly used indentation throughout the codebase, so that's fine.

Reply:
I had a quick go and it touches lots of files outside this commit. Let's resolve whitespaces issues in a separate PR (and fix trailing spaces and tabs too).